### DISCUSSION TITLE
Why the move to _ some methods on token in spacy 3?

### DISCUSSION THREAD

**Discussion by martymcgee3** [explosion/spaCy#9937] [Help: Coding & Implementations] [answered]
I'm following along with a [Youtube](https://www.youtube.com/watch?v=WnGPv6HnBok) tutorial that has great content but is a little dated.  One of the code snippets includes:

```python
def has_golang(doc):
    for t in doc:
        if t.lower_ in ["go", "golang"]:
            if t.pos_ == "NOUN":
                return True 
    return False
```
...which fails because t.lower, somewhat confusingly now returns a number (the hash / vector?), instead of the lower case variant of the token.  `t.lower_` seems to be the new method.

As someone very new to spacy and somewhat inexperienced with Python, I'm wondering what the rationale is for this?  Is it to make vector comparisons work as expected?  Am I right in interpreting the underscore at the end as indicating this method is internal and should not be relied upon?

Thanks for the incredible library, really enjoying it!

**Comment by polm** [comment, marked as answer]
The code in the video ([time link](https://youtu.be/WnGPv6HnBok?t=990)), including as you have pasted it, uses `lower_` with an underscore. This hasn't changed between v2 and v3. 

The underscore methods are for token attributes that are strings. For efficiency purposes they are stored as hashes (integers), and the underscore versions convert the hashes back to strings. If you want to read more about this you can see the [dev docs](https://github.com/explosion/spaCy/blob/master/extra/DEVELOPER_DOCS/StringStore-Vocab.md#:~:text=one%20pattern%20you'll%20see%20a%20lot%20in%20spacy%20apis), though you don't have to know all that to use the API. 

Also, the general convention in Python is that members that **start** with an underscore are for internal use. The spaCy underscore properties can be used freely.

**Reply by martymcgee3** [reply]
Wow, I guess I totally transcribed it incorrectly :man_facepalming:.  Thanks for setting me straight, and the additional info on the hashes and underscore conventions in Spacy.
